tests/tracer/test_encoders.py

```diff
@@ -595,3 +598,71 @@ def test_encoding_invalid_data(data):
         encoder.put(trace)
 
     assert encoder.encode() is None
+
+
+def test_custom_msgpack_encode_multithreaded():
```

Comment:
while this test confirms the behavior and issue we are seeing, it is very dependent on implementation details which can change at any time. For example, the lock is currently in the wrong place, but if it was moved but still wrong, then this test might not fail or become flaky.

the maintenance burden of this test case is likely going to be high, and we'll probably miss a real failure thinking it is just flaky.

Reply:
Yeh apologies for not spotting this earlier. I just happened to realise the lock mess once I moved away from the code and looked it at it recently with different 👀 😞 . I haven't had the chance to look closer to the details but I believe that if we guard all the encoder state that is shared by threads we should be good. The state includes, at the very least, the actual byte buffer and the trace count (same for the string table).